### PR TITLE
Tensor parallel distributed strategy without using deepspeed

### PR DESCRIPTION
**Pull request by kalyanjk**
Tensor parallel by extending GaudiLlamaAttention -> TPGaudiLlamaAttention and GaudiLlamaMLP -> TPGaudiLlamaMLP

use parameter --distributed_strategy="tp" to invoke this code path

code design reference: https://github.com/foundation-model-stack/foundation-model-stack/tree/main

### CODE REVIEW

File: optimum/habana/distributed/serialization.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
+import collections
+import os
+from collections import ChainMap
+from collections.abc import Iterable
+from pathlib import Path
+from typing import Any, Callable, List, Mapping, MutableMapping, Optional, Union
+
+import torch
+
+from optimum.habana.distributed.tp import TPModule
```
Comment (regisss) [inline optimum/habana/distributed/serialization.py:10]:
```suggestion
from .tp import TPModule
```

File: optimum/habana/distributed/serialization.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+    if not len(state_dict):
+        return state_dict
+    adapter = _get_adapter(architecture, source)
+    adapted = adapter(state_dict)
+    return adapted
+
+
+# `models` imports each model class, causing models and adapters to be registered.
+# down here to avoid circular dependencies.
+# from fms import models
```
Comment (regisss) [inline optimum/habana/distributed/serialization.py:96]: Do we need this?

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -20,10 +20,20 @@
     logger,
 )
 
+import copy
+from torch.distributed.distributed_c10d import ProcessGroup
 from ...modeling_attn_mask_utils import (
     _gaudi_prepare_4d_causal_attention_mask,
 )
 
+from optimum.habana.distributed.tp import TPModule
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:29]:
```suggestion
from ....distributed.tp import TPModule
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -20,10 +20,20 @@
...
+from  optimum.habana import distributed
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:30]:
```suggestion
from .... import distributed
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -20,10 +20,20 @@
...
+from optimum.habana.distributed.tensorparallel import (
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:31]:
```suggestion
from ....distributed.tensor_parallel import (
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -20,10 +20,20 @@
...
+    reduce_from_tensor_model_parallel_region,
+)
+
+from optimum.habana.distributed.strategy import DistributedStrategy
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:35]:
```suggestion
from ....distributed.strategy import DistributedStrategy
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 7]
```diff
@@ -20,10 +20,20 @@
...
+from optimum.habana.distributed.strategy import  NoOpStrategy
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:36]:
```suggestion
from ....distributed.strategy import NoOpStrategy
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 8]
```diff
@@ -717,10 +855,16 @@ def __init__(self, config: LlamaConfig):
         self.padding_idx = config.pad_token_id
         self.vocab_size = config.vocab_size
 
+        self.distributed_strategy = config.distributed_strategy
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:858]: You need to add `distributed_strategy` to `LlamaConfig` here: https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/llama/configuration_llama.py
You can follow what is done for `mlp_bias` and `fused_qkv`.

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 9]
```diff
@@ -717,10 +855,16 @@ def __init__(self, config: LlamaConfig):
...
+        config.distributed_strategy = None
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:859]: Why is this needed?

**Review by regisss — COMMENTED** [review]
I left a few comments. Additionally, can you:
- run `make style`?
- add an example command in the README of the text-generation example?
- add a test for it in https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py?
- add a link to the original implementation in all files that are inspired from it?
- check if there are any copyrights to cite?

**Comment by kalyanjk** [comment]
> I left a few comments. Additionally, can you:
> 
> * run `make style`?
> * add an example command in the README of the text-generation example?
> * add a test for it in https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py?
> * add a link to the original implementation in all files that are inspired from it?
> * check if there are any copyrights to cite?

Done

File: examples/text-generation/utils.py [hunk 10]
```diff
@@ -245,6 +245,102 @@ def setup_model(args, model_dtype, model_kwargs, logger):
...
+            return tp_wrapping.apply_tp(block, layer, self.group)
+
+        def __getstate__(self):
+            state = self.__dict__.copy()
+            state['group'] = None  # Remove ProcessGroup from state
+            return state
+
+        def __setstate__(self, state):
+            self.__dict__.update(state)
+            self.group = None  # Restore to default state or reinitialize
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply examples/text-generation/utils.py]: Done

File: examples/text-generation/utils.py [hunk 11]
```diff
@@ -245,6 +245,102 @@ def setup_model(args, model_dtype, model_kwargs, logger):
...
+    logger.info("Multi-device run.")
+
+    assert args.assistant_model is None, "Assistant model must be None"
+    
+    from torch import distributed as dist
+    if args.device == 'hpu':
+        import habana_frameworks.torch.distributed.hccl
+        dist.init_process_group(backend='hccl')
+    else:
+        dist.init_process_group()
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply examples/text-generation/utils.py:265]: Done

File: examples/text-generation/utils.py [hunk 12]
```diff
@@ -245,6 +245,102 @@ def setup_model(args, model_dtype, model_kwargs, logger):
...
+        model = setup_quantization(model, args)
+    
+    model = model.eval().to(args.device)
+
+    if args.use_hpu_graphs:
+        from habana_frameworks.torch.hpu import wrap_in_hpu_graph
+
+        if check_habana_frameworks_version("1.13.0") and model.config.model_type == "falcon":
+            model = wrap_in_hpu_graph(model, hash_with_views=False)
+        else:
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply examples/text-generation/utils.py:309]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 6 above)
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 7 above)
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Done

File: optimum/habana/distributed/tp_wrapping.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+from optimum.habana.transformers.models.llama.modeling_llama import (
+    GaudiLlamaMLP, 
+    TPGaudiLlamaMLP,
+    GaudiLlamaAttention,
+    TPGaudiLlamaAttention
+)
+
+# this probably belongs somewhere else but can't go in fms.distribtued b/c
+# circular dependency.
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply optimum/habana/distributed/tp_wrapping.py:30]: Done

File: optimum/habana/distributed/tp.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
+import itertools
+from abc import ABCMeta, abstractmethod
+from typing import List, Type
+
+import torch
+import torch.nn as nn
+from torch.distributed.distributed_c10d import ProcessGroup
+
+from optimum.habana.distributed.tensorparallel import (
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply optimum/habana/distributed/tp.py:26]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 5 above)
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Done

File: optimum/habana/distributed/tp_wrapping.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
+import os
+from torch import nn
+from torch.distributed.distributed_c10d import ProcessGroup
+
+from optimum.habana.transformers.models.llama.modeling_llama import (
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply optimum/habana/distributed/tp_wrapping.py:21]: Done

File: optimum/habana/distributed/serialization.py (hunk 1 above)
Comment (kalyanjk) [reply optimum/habana/distributed/serialization.py:27]: Done

File: optimum/habana/distributed/serialization.py (hunk 2 above)
Comment (kalyanjk) [reply optimum/habana/distributed/serialization.py:103]: Removed

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 3 above)
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py:26]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 4 above)
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Done

File: optimum/habana/distributed/tensorparallel.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
+# mypy: disable-error-code="method-assign,misc"
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply optimum/habana/distributed/tensorparallel.py:18]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 8 above)
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 9 above)
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Removed

**Comment by kalyanjk** [comment]
@regisss Addressed all your comments. Please review the changes. Thank you!

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 8 above)
Comment (regisss) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Hmm I don't see this change

Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: I had removed this change  self.distributed_strategy = config.distributed_strategy
so it thought it was not necessary since i dint see it getting used any where.
I have updated the  LlamaConfig now.

Comment (regisss) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Yeah, it may not be absolutely necessary but I feel like it's cleaner that way.

File: tests/test_text_generation_example.py [hunk 17]
```diff
@@ -71,6 +71,9 @@
         "torch_compile_distributed": [
             ("meta-llama/Llama-2-7b-hf", 39.72973199515235),
         ],
+        "distributed_tp": [
+            ("/mnt/weka/data/pytorch//llama2/Llama-2-7b-hf/", 1856.8140409694543),
```
Comment (regisss) [inline tests/test_text_generation_example.py:75]:
```suggestion
            ("meta-llama/Llama-2-7b-hf", 1856.8140409694543),
```
This is your local path, let's switch to the hub id.

File: tests/test_text_generation_example.py [hunk 18]
```diff
@@ -142,11 +152,13 @@ def _test_text_generation(
...
         command += ["--use_flash_attention", "--flash_attention_causal_mask"]
 
-    if reuse_cache or torch_compile:
+    if (reuse_cache or torch_compile) and not distributed_strategy == "tp":
         command += ["--reuse_cache"]
 
     if torch_compile:
         command += ["--torch_compile"]
+        command += ["--use_flash_attention"]
+        command += ["--flash_attention_recompute"]
```
Comment (regisss) [inline tests/test_text_generation_example.py:161]: Let's add these only if `distributed_strategy == "tp"`

File: examples/text-generation/README.md [hunk 19]
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
 
 You will also need to add `--torch_compile` in your command.
 
+### Running with Tesor parallel strategy
```
Comment (regisss) [inline examples/text-generation/README.md:267]:
```suggestion
### Running with tensor-parallel strategy
```

File: examples/text-generation/README.md [hunk 20]
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
...
+#### Attribution
+
+This repository includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
```
Comment (regisss) [inline examples/text-generation/README.md:270]:
```suggestion
> [!NOTE]  
> This strategy includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
```

File: examples/text-generation/README.md [hunk 21]
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
 
 You will also need to add `--torch_compile` in your command.
 
+### Running with Tesor parallel strategy
+#### Attribution
```
Comment (regisss) [inline examples/text-generation/README.md:268]: I think you can remove that line, let's put it in a "box" as suggested below

File: examples/text-generation/README.md [hunk 22]
```diff
@@ -264,6 +264,35 @@ set the following environment variables before running the command: `PT_ENABLE_I
...
+
+This repository includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
+
+torch.compile with tensor parallel strategy is an experimental feature. It has not been validated for all models. To enable
```
Comment (regisss) [inline examples/text-generation/README.md:272]:
```suggestion
> [!WARNING]  
> torch.compile with tensor parallel strategy is an experimental feature. It has not been validated for all models.

To enable...
```

**Review by regisss — COMMENTED** [review]
Can you update your main branch and merge it into this PR? To have the whole CI working again.

**Comment by libinta** [comment]
@kalyanjk can you update based on review comments?

File: tests/test_text_generation_example.py (hunk 17 above)
Comment (kalyanjk) [reply tests/test_text_generation_example.py]: > This is your local path, let's switch to the hub id.

Done

File: tests/test_text_generation_example.py (hunk 18 above)
Comment (kalyanjk) [reply tests/test_text_generation_example.py]: Done

File: examples/text-generation/README.md (hunk 19 above)
Comment (kalyanjk) [reply examples/text-generation/README.md]: Done

File: examples/text-generation/README.md (hunk 22 above)
Comment (kalyanjk) [reply examples/text-generation/README.md]: Done, Added Note and Warning in box

File: examples/text-generation/README.md (hunk 20 above)
Comment (kalyanjk) [reply examples/text-generation/README.md]: Done

File: examples/text-generation/README.md (hunk 21 above)
Comment (kalyanjk) [reply examples/text-generation/README.md]: I have added in the box, but not sure if this syntax had to be preserved [!WARNING]

**Comment by kalyanjk** [comment]
> @kalyanjk can you update based on review comments?

@libinta, I have addressed all.  Anything i have missed?

**Review by regisss — COMMENTED** [review]
I left a few more comments to address.
Also, the test fails on my instance with this error:
```
Traceback (most recent call last):                                                                                                           
  File "/root/workspace/fork/examples/text-generation/run_generation.py", line 674, in <module>                                              
    main()                                                                                                                                   
  File "/root/workspace/fork/examples/text-generation/run_generation.py", line 317, in main                                                  
    model, assistant_model, tokenizer, generation_config = initialize_model(args, logger)                                                    
  File "/root/workspace/fork/examples/text-generation/utils.py", line 592, in initialize_model                                               
    else setup_distributed_model_tp(args, model_dtype, model_kwargs, logger)                                                                 
  File "/root/workspace/fork/examples/text-generation/utils.py", line 281, in setup_distributed_model_tp                                     
    lazy_sd = serialization.load_state_dict(                                                                                                 
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/distributed/serialization.py", line 191, in load_state_dict                   
    assert len(checkpoints) > 0, f"Can't find the requested checkpoint data at {model_path}"                                                 
AssertionError: Can't find the requested checkpoint data at meta-llama/Llama-2-7b-hf
```
Any idea about what's going on? It seems like a serialization issue. Or is it because it requires Synapse 1.17? I'm running 1.16.

**Comment by regisss** [comment]
Please run `make style` too

File: examples/text-generation/README.md [hunk 23]
```diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
...
+WARNING: torch.compile with tensor parallel strategy is an experimental feature. It has not been validated for all models.
+```
+To enable torch.compile with tensor parallel strategy, please set the following environment variables before running the
+command: `PT_ENABLE_INT64_SUPPORT=1` and `PT_HPU_LAZY_MODE=0`. This will enable tensor parallel strategy without deepspeed.
+
+You will also need to add `--torch_compile` and `--distributed_strategy="tp"` in your command.
+
+Here is an example:
+```bash
+python ../gaudi_spawn.py  --world_size 8 run_generation.py \
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply examples/text-generation/README.md]: updated

File: examples/text-generation/README.md [hunk 24]
```diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
...
+
+```bash
+NOTE: This strategy includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
+```
+
+```bash
+WARNING: torch.compile with tensor parallel strategy is an experimental feature. It has not been validated for all models.
+```
+To enable torch.compile with tensor parallel strategy, please set the following environment variables before running the
+command: `PT_ENABLE_INT64_SUPPORT=1` and `PT_HPU_LAZY_MODE=0`. This will enable tensor parallel strategy without deepspeed.
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply examples/text-generation/README.md]: updated with the suggested format

File: examples/text-generation/README.md [hunk 25]
```diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
 
 You will also need to add `--torch_compile` in your command.
 
+### Running with tesor-parallel strategy
+
+```bash
+NOTE: This strategy includes code from the [foundation-model-stack](https://github.com/foundation-model-stack/foundation-model-stack) repository, which is licensed under the Apache License 2.0. See the `LICENSE` file for more details.
+```
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply examples/text-generation/README.md]: updated with the suggested format

File: examples/text-generation/README.md [hunk 26]
```diff
@@ -264,6 +264,38 @@ set the following environment variables before running the command: `PT_ENABLE_I
 
 You will also need to add `--torch_compile` in your command.
 
+### Running with tesor-parallel strategy
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply examples/text-generation/README.md]: Done

**Comment by kalyanjk** [comment]
> I left a few more comments to address. Also, the test fails on my instance with this error:
> 
> ```
> Traceback (most recent call last):                                                                                                           
>   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 674, in <module>                                              
>     main()                                                                                                                                   
>   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 317, in main                                                  
>     model, assistant_model, tokenizer, generation_config = initialize_model(args, logger)                                                    
>   File "/root/workspace/fork/examples/text-generation/utils.py", line 592, in initialize_model                                               
>     else setup_distributed_model_tp(args, model_dtype, model_kwargs, logger)                                                                 
>   File "/root/workspace/fork/examples/text-generation/utils.py", line 281, in setup_distributed_model_tp                                     
>     lazy_sd = serialization.load_state_dict(                                                                                                 
>   File "/usr/local/lib/python3.10/dist-packages/optimum/habana/distributed/serialization.py", line 191, in load_state_dict                   
>     assert len(checkpoints) > 0, f"Can't find the requested checkpoint data at {model_path}"                                                 
> AssertionError: Can't find the requested checkpoint data at meta-llama/Llama-2-7b-hf
> ```
> 
> Any idea about what's going on? It seems like a serialization issue. Or is it because it requires Synapse 1.17? I'm running 1.16.

Can you provide with absolute path for meta-llama/Llama-2-7b-hf.  All my testing is on 1.17, I will verify on 1.16 and update.

**Comment by kalyanjk** [comment]
> > I left a few more comments to address. Also, the test fails on my instance with this error:
> > ```
> > Traceback (most recent call last):                                                                                                           
> >   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 674, in <module>                                              
> >     main()                                                                                                                                   
> >   File "/root/workspace/fork/examples/text-generation/run_generation.py", line 317, in main                                                  
> >     model, assistant_model, tokenizer, generation_config = initialize_model(args, logger)                                                    
> >   File "/root/workspace/fork/examples/text-generation/utils.py", line 592, in initialize_model                                               
> >     else setup_distributed_model_tp(args, model_dtype, model_kwargs, logger)                                                                 
> >   File "/root/workspace/fork/examples/text-generation/utils.py", line 281, in setup_distributed_model_tp                                     
> >     lazy_sd = serialization.load_state_dict(                                                                                                 
> >   File "/usr/local/lib/python3.10/dist-packages/optimum/habana/distributed/serialization.py", line 191, in load_state_dict                   
> >     assert len(checkpoints) > 0, f"Can't find the requested checkpoint data at {model_path}"                                                 
> > AssertionError: Can't find the requested checkpoint data at meta-llama/Llama-2-7b-hf
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any idea about what's going on? It seems like a serialization issue. Or is it because it requires Synapse 1.17? I'm running 1.16.
> 
> Can you provide with absolute path for meta-llama/Llama-2-7b-hf. All my testing is on 1.17, I will verify on 1.16 and update.

@regisss successfully verified the sanity test for the 1.16 release using both the 7b and 70b models. Everything is working fine.

**Comment by regisss** [comment]
There is no absolute path, this is the hub model id and I really think this use case should work as not everybody has the models stored locally. If the absolute path to the model is needed, there should be some code to find the model in the Transformers cache. You can get the default path to cache with:
```
from huggingface_hub.constants import HF_HUB_CACHE
```
More information about the structure of the cache here: https://huggingface.co/docs/huggingface_hub/v0.24.2/en/guides/manage-cache#understand-caching

Also, I see I forgot to mention it, can you replace the arg `distributed_strategy` by `parallelism_strategy` or something similar everywhere please? We already have `distribution_strategy` defined [here](https://github.com/huggingface/optimum-habana/blob/a4851cde758fe8e50044f43d1104cbd2e77b1ff6/optimum/habana/transformers/training_args.py#L169) and it would add a lot of confusion to have both `distributed_strategy` and `distribution_strategy`. Sorry for noticing this now, I thought I had commented on it before.

**Comment by kalyanjk** [comment]
> There is no absolute path, this is the hub model id and I really think this use case should work as not everybody has the models stored locally. If the absolute path to the model is needed, there should be some code to find the model in the Transformers cache. You can get the default path to cache with:
> 
> ```
> from huggingface_hub.constants import HF_HUB_CACHE
> ```
> 
> More information about the structure of the cache here: https://huggingface.co/docs/huggingface_hub/v0.24.2/en/guides/manage-cache#understand-caching
> 
> Also, I see I forgot to mention it, can you replace the arg `distributed_strategy` by `parallelism_strategy` or something similar everywhere please? We already have `distribution_strategy` defined [here](https://github.com/huggingface/optimum-habana/blob/a4851cde758fe8e50044f43d1104cbd2e77b1ff6/optimum/habana/transformers/training_args.py#L169) and it would add a lot of confusion to have both `distributed_strategy` and `distribution_strategy`. Sorry for noticing this now, I thought I had commented on it before.

Updated : renamed the distributed_strategy to parallel_strategy.  
In process : Is there a way i can test the relative path for model_name - > meta-llama/Llama-2-7b-hf.

**Comment by kalyanjk** [comment]
> > There is no absolute path, this is the hub model id and I really think this use case should work as not everybody has the models stored locally. If the absolute path to the model is needed, there should be some code to find the model in the Transformers cache. You can get the default path to cache with:
> > ```
> > from huggingface_hub.constants import HF_HUB_CACHE
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > More information about the structure of the cache here: https://huggingface.co/docs/huggingface_hub/v0.24.2/en/guides/manage-cache#understand-caching
> > Also, I see I forgot to mention it, can you replace the arg `distributed_strategy` by `parallelism_strategy` or something similar everywhere please? We already have `distribution_strategy` defined [here](https://github.com/huggingface/optimum-habana/blob/a4851cde758fe8e50044f43d1104cbd2e77b1ff6/optimum/habana/transformers/training_args.py#L169) and it would add a lot of confusion to have both `distributed_strategy` and `distribution_strategy`. Sorry for noticing this now, I thought I had commented on it before.
> 
> Updated : renamed the distributed_strategy to parallel_strategy. In process : Is there a way i can test the relative path for model_name - > meta-llama/Llama-2-7b-hf.

Updated cache_dir setting for  parallel_strategy = tp

@regisss  can you please verify if you are able to load the data now

File: tests/test_text_generation_example.py [hunk 27]
```diff
@@ -71,6 +71,9 @@
         "torch_compile_distributed": [
             ("meta-llama/Llama-2-7b-hf", 39.72973199515235),
         ],
+        "distributed_tp": [
+            ("meta-llama/Llama-2-7b-hf", 1856.8140409694543),
```
Comment (regisss) [inline tests/test_text_generation_example.py:75]: I get 1345.2369318328463 on my Gaudi2 instance with 1.16. Could you check if you get the same? I'm assuming the gap comes from 1.17, but I'm really not sure about it.

**Review by regisss — COMMENTED** [review]
Thanks for the changes, it looks good to me!
One last thing, as written in the comment below, the test fails on my instance because the throughput I get is too low. Maybe due to a different version of Synapse?

File: tests/test_text_generation_example.py (hunk 27 above)
Comment (kalyanjk) [reply tests/test_text_generation_example.py]: @regiss I observed a performance number of 1361.2050636537415 on the 1.16 release, which is close what you have observed. I have updated the expected performance number accordingly.

**Review by regisss — APPROVED** [review]
